### ISSUE TITLE
GestureRecognizers are lost when using GlideX

### ISSUE THREAD

**Issue by bijington** [jonathanpeppers/glidex#2]
In our application we add GestureRecognizers to handle tap events on images but it seems that by enabling Glide it then stops those working.

Just a simple example of what we have configured.

    <Image Source="{Binding Thumbnail.Value}"
           Grid.Row="3"
           IsVisible="{Binding ShowAttachment}"
           Aspect="AspectFit"
           BackgroundColor="Transparent"
           HorizontalOptions="Fill"
           HeightRequest="300">
        <Image.GestureRecognizers>
            <TapGestureRecognizer Command="{Binding ShowAttachmentCommand}"
                                  NumberOfTapsRequired="1" />
        </Image.GestureRecognizers>
    </Image>

I do plan to try subclassing `ImageRenderer` as mentioned in the readme but I thought it worth raising this.

**Comment by jonathanpeppers** [comment]
If you take glidex's `ImageRenderer` as a starting point and get it to work, send a PR!

Here is XF's source, I used as a starting point myself: https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.Android/FastRenderers/ImageRenderer.cs

I am working on a PR for Xamarin.Form itself, where glidex shouldn't need custom renderers anymore. We would be able to get by just using `IImageSourceHandler`. I think that would be a better fix for this issue, because glidex doesn't really want to be in the business of handling gestures...

**Comment by bijington** [comment]
@jonathanpeppers thanks for the pointer. I would agree that it makes sense to retain the gesture handling within Xamarin Forms. 

Out of interest (and I appreciate this is probably a difficult question to answer) do you have any indication of when your PR is likely to make it in to a release assuming all goes well? I would be fine with waiting if it isn't going to be a long time.

**Comment by jonathanpeppers** [comment]
Even if merged quickly, I bet it wouldn’t get released until Xamarin.Forms 3.1 or 3.2 (but really I have no idea).

If you need gestures soon, I would look into making a custom renderer.

**Comment by bijington** [comment]
@jonathanpeppers I suspected as much. Thanks I will likely head down the route of a custom renderer.

I did attempt a quick investigation however I rather embarrassingly struggled to get my new custom renderer to override your ImageRenderer. I must have missed something though so I will look back in to it.

**Comment by jonathanpeppers** [comment]
You should be able to copy what I did here: https://github.com/jonathanpeppers/glidex/blob/065f02664e8034207aa9cc2afb4e3f83f14c9e42/glidex.forms/ImageRenderer.cs#L9

Although... I'm not sure what would make Xamarin.Forms pick _your_ renderer over the Glide one. I don't know how that ordering works...

Maybe you should clone this repo and try making the change in Glide? I'll help you through a pull request and get a release on NuGet quickly if everything works.

**Comment by bijington** [comment]
@jonathanpeppers that is what I had initially tried but it didn't work and I couldn't even hit a breakpoint in my new renderers constructor. Perhaps I needed to clean and redeploy.

I will have a further play and see, failing that I will look to clone the repo and make the change like you suggested.

**Comment by jonathanpeppers** [comment]
@bijington don't try doing this one yourself...

I can't figure out a way to make it work currently, since glidex's renderer is based on the "fast" `ImageRenderer`.

To get gestures to work, it requires usage of internal classes: https://github.com/xamarin/Xamarin.Forms/blob/d72e8dd43bde3c57d45f1a95f7a8b52d044d93a5/Xamarin.Forms.Platform.Android/FastRenderers/ImageRenderer.cs#L79

https://github.com/xamarin/Xamarin.Forms/blob/d72e8dd43bde3c57d45f1a95f7a8b52d044d93a5/Xamarin.Forms.Platform.Android/FastRenderers/VisualElementRenderer.cs

Gestures work on the older "slow" `ImageRenderer` because it subclasses `ViewRenderer`...

**Comment by bijington** [comment]
@jonathanpeppers thanks for the update. With your planned PR for Xamarin Forms will this get around the issue? If so then I am happy to wait.

**Comment by jonathanpeppers** [comment]
Yeah the PR would mean glidex would not have to make a custom renderer.

**Comment by bijington** [comment]
That is fine with me then. Feel free to close this down.

**Comment by BillyMartin1964** [comment]
@jonathanpeppers 
Even after uninstalling GlideX.forms, the gesture recognizers fail to work. What type of Permanent change could it have done to my Xamarin Forms Android project and how can I get it back?

**Comment by jonathanpeppers** [comment]
@BillyMartin1964 uninstall the app?

Perhaps this is a bug in Fast Deployment, I’ll make a note to test this.

**Comment by BillyMartin1964** [comment]
@jonathanpeppers , Yes. I unistalled the nuget package and commented out the init in main activity, but the gesture recognizers still wouldn't work. Fortunately I had backed up just prior to trying glidex! Lol Please let me know when it's stable. It sounds like a great package.

**Comment by jonathanpeppers** [comment]
So this got merged in Xamarin.Forms master: https://github.com/xamarin/Xamarin.Forms/pull/3045

It should be in the _next_ preview (after the 3.2.x series).

When that preview comes out, I'll make a preview package of `glidex.forms` using it. And gestures should work again!

**Comment by BillyMartin1964** [comment]
Gestures are pretty freaking important don't you think? Can't believe you want to wait on something so important.

**Comment by jonathanpeppers** [comment]
@BillyMartin1964 keep in mind this is a project I work on in my copious spare time. I will accept pull requests, no problem! I would even help you get it merged!

What do you mean “wait”? The XF pull request above is mine, and the path forward. How do you recommend I proceed?

**Comment by BillyMartin1964** [comment]
I meant no offense to YOU. I just meant that gestures are a big part of images. Microsoft made it a point to brag about how much GlideX speeded up Xamarin Forms Android. Microsoft should throw some of their vast money at it and give you help!

**Comment by jonathanpeppers** [comment]
@BillyMartin1964 Ok, good to know.

I think this project has done what it was meant to do:
1. Show how poor `Image` performance is on Android in Xamarin.Forms, with concrete examples.
1. Show a clear, Android-native route to improving `Image` performance.
1. Provide a NuGet package for people on the bleeding edge to try out in their Xamarin.Forms applications.

I think Glide (and the concepts here) will make their way into Xamarin.Forms itself--but I'm not actually a part of that team, so I don't know how long it will take.

**Comment by bijington** [comment]
@jonathanpeppers excellent work on the PR! I look forward to trying it out in a not so distant XF release.

**Comment by jonathanpeppers** [comment]
Guys go upvote this https://github.com/xamarin/Xamarin.Forms/issues/3577

It is happening!!!

**Comment by BillyMartin1964** [comment]
@jonathanpeppers  I'm having a problem when I hit the android back button and then resume the app:

Java.Lang.IllegalArgumentException: You cannot start a load for a destroyed activity

Everything I google says it's been a glide problem for a while, but after commenting it out in the main activity, uninstalling the app, and rebooting the phone, it still happens when I debug it again. I didn't try uninstalling the package, though.  I upvoted Glide for XF, but it has too be stable.

**Comment by jonathanpeppers** [comment]
@BillyMartin1964 can you open a new issue with the full error message you are getting? Post a link to what you found searching the web?

**Comment by BillyMartin1964** [comment]
@jonathanpeppers Yeah, I'll do that. I just uninstalled Glidex.Forms and my app works fine. Definitely Glide issue. Where do you want me to open the issue?

**Comment by BillyMartin1964** [comment]
@jonathanpeppers , It's done. 

[](https://github.com/jonathanpeppers/glidex/issues/5)

**Comment by jonathanpeppers** [comment]
@bijington or @BillyMartin1964 I released a preview NuGet based on the Xamarin.Forms 3.3 preview: https://www.nuget.org/packages/glidex.forms/1.0.3-pre

There are no longer any renderers, so gestures should work. Could you try it out and let me know?

When Xamarin.Forms goes stable, I'd like `glidex.forms` to go stable around the same time.

Closing this, as I think the prerelease solves this problem. We can reopen/open a new issue if something new is broken.
